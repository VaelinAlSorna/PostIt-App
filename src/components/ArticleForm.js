import React from 'react';
import moment from 'moment';

class ArticleForm extends React.Component {
    state = {
        title: '',
        text: '',
        createdAt: moment().format('D. M. Y')
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({
            title
        }))
    }
    onTextChange = (e) => {
        const text = e.target.value;
        this.setState(() => ({
            text
        }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit(this.state)
    }
    render () {
       return ( 
        <div> 
            <form onSubmit={this.onSubmit}> 
                <input 
                    onChange={this.onTitleChange} 
                    placeholder="Title" 
                    type="text" 
                    value={this.state.title}
                />
                <textarea 
                    onChange={this.onTextChange}
                    placeholder="Add text for your article"
                    value={this.state.text}
                >
                </textarea>
                <input type="submit" value="Submit" />
            </form>
        </div>
       );
    }
}

export default ArticleForm;