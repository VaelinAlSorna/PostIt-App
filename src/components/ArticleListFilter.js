import React from 'react';
import { connect } from 'react-redux';
import { setYourArticlesFilter, removeYourArticlesFilter, setTextFilter } from '../actions/filterActions';

export class ArticleListFilter extends React.Component {
    handleSeeAll = () => {
            this.props.removeYourArticlesFilter();
            this.props.setTextFilter('');
    }
    handleSeeYours = () => {
        this.props.setYourArticlesFilter();
    }
    disablingBtnSeeAll = () => {
        if(this.props.filter.yourArticles || !!this.props.filter.text) {
            return false;
        } else {
            return true;
        }
    }
    disablingBtnSeeYours = () => {
        if(this.props.user.name && !this.props.filter.yourArticles) {
            return false;
        } else {
            return true;
        }
    }
    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }
    onSubmit = (e) => {
        //This is here in order to prevent user on mobile from sunmiting form
        e.preventDefault();
    }
    
    render () { 
        return  ( 
            <div className="filter">
                <form onSubmit={this.onSubmit} className="filter__group filter-form"> 
                    <input
                        className="filter-form__input"
                        onChange={this.onTextFilterChange} 
                        type="text" 
                        placeholder="Search Article By Name"
                        value={this.props.filter.text}
                    />
                </form>
                <button className="filter__group btn btn--green" onClick={this.handleSeeAll} disabled={this.disablingBtnSeeAll()}>See All Articles</button>
                <button className="filter__group btn btn--green" onClick={this.handleSeeYours} disabled={this.disablingBtnSeeYours()}>See Your Articles</button>
            </div>                
        )
    };
}

const mapStateToProps = (state) => ({
    filter: state.filter,
    user: state.auth
});

const mapDispatchToProps = (dispatch) => ({
    setYourArticlesFilter: () => dispatch(setYourArticlesFilter()),
    removeYourArticlesFilter: () => dispatch(removeYourArticlesFilter()),
    setTextFilter: (text) => dispatch(setTextFilter(text))
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleListFilter);

