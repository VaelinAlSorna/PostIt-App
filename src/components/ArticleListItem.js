import React from 'react';
//import { IoIosHeart, IoIosChatboxes } from "react-icons/io";
import { Link } from 'react-router-dom';


const ArticleListItem = ({ _id, title, createdAt, creator, comments, likes}) => (
    <Link to={`view/${_id}`}>
        <h3>{title}</h3>
        <p>Posted by: {creator} at: {createdAt}</p>
        {/*<p> <IoIosHeart /> {likes.length}</p>*/}
        {/*<p><IoIosChatboxes /> { comments.length }</p>*/}
    </Link>
);

export default ArticleListItem;