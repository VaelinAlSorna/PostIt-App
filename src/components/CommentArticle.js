import React from 'react';
import uuid from 'uuid';
import moment from 'moment';

class CommentArticle extends React.Component {
    state = {
        comment: '',
        createdAt: moment().format('D. M. Y'),
        error: ''
    };
    onTextareaChange = (e) => {
        const comment = e.target.value;
        this.setState(() => ({
            comment
        }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (this.state.comment.trim().length < 3) {
            this.setState(() => ({
                error: 'Yout comment must contain at least 3 characters.'
            }));
        } else {
            const id = uuid();
            this.props.onSubmit({...this.state, id});
            this.setState(() =>  ({
                comment: '',
                error: ''
            }));
        }
    }
    render () {
        return (
            <form onSubmit={this.onSubmit}>
                { this.state.error && <p>{this.state.error}</p>}
                <textarea 
                    onChange={this.onTextareaChange} 
                    placeholder="What are your thoughts ?"
                    value={this.state.comment}
                >
                </textarea>
                <input type="submit" value="Comment" />
            </form>
        );
    };
};

export default CommentArticle;