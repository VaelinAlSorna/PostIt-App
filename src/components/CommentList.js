import React from 'react';
import CommentListItem from './CommentListItem';

const CommentList = (props) => (
         <ol>
            {
                props.comments.map((comment) => <CommentListItem 
                                                userId={props.userId}
                                                onClickDeleteComment={props.onClickDeleteComment} 
                                                key={comment._id} 
                                                {...comment}/>
                                            )
            }
        </ol>
)

export default CommentList;