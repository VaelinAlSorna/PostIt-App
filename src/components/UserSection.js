import React from 'react';
import { connect } from 'react-redux'
import LoginForm from './LoginForm';
import  SignUpForm from './SingUpForm';
import LoggedIn from './LoggedIn';
import { history } from '../routers/AppRouter';
import { startSignUp, startLogin, startLogout } from '../actions/authActions';
import { setYourArticlesFilter, removeYourArticlesFilter } from '../actions/filterActions';

export class UserSection extends React.Component {
    state = {
        //this determines on which form I am currently
        login: true
    };
    changeForm = () => {
        this.state.login ? this.setState(() => ({ login: false})) : this.setState(() => ({ login: true}));
    };
    onLogout = () => {
        //this here ensures that user gets always redirected to the login section when logout(and not to the signup section)
        this.setState(() => ({
            login: true
        }));
        this.props.startLogout(this.props.user.token);
        this.props.removeYourArticlesFilter(); //this removes filter so when log out user can actually see articles
        //this  redirects the user in case he logs out while creating or edititng article
        if(window.location.href.includes('edit') || window.location.href.includes('create')){
            history.push('/');
        }
    };
    onSubmitLogin = (userData) => {
        this.props.startLogin(userData)
    };
    onSubmitSignUp = (userData) => {
        this.props.startSignUp(userData);
    };
    renderCorrectPart = () => {
        // this checks if user exists
        if(!!this.props.user.name) {
            return <LoggedIn filter={this.props.filter} handleFilter={this.handleFilter} name={this.props.user.name} onLogout={this.onLogout} />
        } else if (this.state.login) {
            return <LoginForm authFailed={this.props.authFailed} changeForm={this.changeForm} onSubmit={this.onSubmitLogin}/>
        } else {
            return <SignUpForm authFailed={this.props.authFailed} changeForm={this.changeForm} onSubmit={this.onSubmitSignUp}/>
        }
    };
    handleFilter = () => {
        if (this.props.filter){
            this.props.removeYourArticlesFilter();
            history.push('/');
        }else {
            this.props.setYourArticlesFilter();
            history.push('/');
        }
    }
    render () {
        return (
            <div>
                { this.props.networkError ? <p>We are sorry but some error probably occured</p> : this.renderCorrectPart()  }
            </div>
        );
    };
};
const mapStateToProps = (state) => ({
    user: state.auth,
    authFailed: state.auth.authFailed,
    filter: state.filter,
    networkError: state.networkError
});

const mapDispatchToProps = (dispatch) => ({
    startSignUp: (userData) => dispatch(startSignUp(userData)),
    startLogin: (userData) => dispatch(startLogin(userData)),
    startLogout: (userToken) => dispatch(startLogout(userToken)),
    setYourArticlesFilter: () => dispatch(setYourArticlesFilter()),
    removeYourArticlesFilter: () => dispatch(removeYourArticlesFilter())
});

export default connect(mapStateToProps, mapDispatchToProps)(UserSection);