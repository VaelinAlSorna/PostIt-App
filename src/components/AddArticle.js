import React from 'react';
import { connect } from 'react-redux'
import ArticleForm from './ArticleForm';
import { addArticle } from '../actions/articlesActions'


class AddArticle extends React.Component {
    onSubmit = (article) => {
        this.props.dispatch(addArticle(article))
        this.props.history.push('/');
    }
    render () {
        return ( 
        <div>
            <ArticleForm onSubmit={this.onSubmit} />
        </div>
        );
    }
}

export default connect(undefined)(AddArticle);

