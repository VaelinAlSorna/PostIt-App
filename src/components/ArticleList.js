import React from 'react';
import { connect } from 'react-redux';
import ArticleListItem from './ArticleListItem';
import yourArticlesFilter from '../selectors/yourArticlesFilter';

export const ArticleList = (props) => (
    <div>
        {
            !!props.networkError 
            ? 
            (<h1>Server is down please try reconnect later </h1>)
            :
            (<ol>
                {
                    (props.articles.length === 0 && !!props.filter ) 
                    ? <p> You haven't created any articles yet. </p>
                    :props.articles.map((article) => <ArticleListItem key={article._id} {...article} />)
                }
            </ol>)
        }
    </div>
);

const mapStateToProps = (state) => ({
    articles: state.filter ? yourArticlesFilter(state.articles, state.auth.id) : state.articles,
    filter: state.filter,
    networkError: state.networkError
});

export default connect(mapStateToProps)(ArticleList);