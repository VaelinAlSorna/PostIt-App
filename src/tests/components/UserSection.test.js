import React from 'react';
import { shallow } from 'enzyme';
import { UserSection } from '../../components/UserSection';
import SignUpForm from '../../components/SingUpForm';
import user from '../fixtures/usersFixtures';

let startSignUp, 
    startLogin, 
    startLogout, 
    setYourArticlesFilter, 
    removeYourArticlesFilter,
    history,
    wrapper;

beforeEach(() => {
    startSignUp = jest.fn();
    startLogin =jest.fn();
    startLogout = jest.fn();
    setYourArticlesFilter = jest.fn();
    removeYourArticlesFilter = jest.fn();
    history= { push: jest.fn() }
    wrapper = shallow(<UserSection 
        startSignUp={startSignUp}
        startLogin={startLogin}
        startLogout={startLogout}
        setYourArticlesFilter={setYourArticlesFilter}
        removeYourArticlesFilter={removeYourArticlesFilter}
        user={{}}
        authFailed={undefined}
        filter={null}
        networkError={false}
        history={history}
                    />)
});

test('should render UserSection when loginIn', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render UserSection when signingUp', () => {
    wrapper.setState({ login: false });
    expect(wrapper).toMatchSnapshot();
});

test('should render UserSection when user is loggedIn', () => {
    wrapper = shallow(<UserSection 
        startSignUp={startSignUp}
        startLogin={startLogin}
        startLogout={startLogout}
        setYourArticlesFilter={setYourArticlesFilter}
        removeYourArticlesFilter={removeYourArticlesFilter}
        user={user} //<-- this is diffrent from befaoreach shallow
        authFailed={undefined}
        filter={null}
        networkError={false}
        history={history}
                    />)
    expect(wrapper).toMatchSnapshot();
});

test('should call startLogin onSubmitLogin when I am on on LoginForm', () => {
    wrapper.find('LoginForm').prop('onSubmit')(user);
    expect(startLogin).toHaveBeenLastCalledWith(user);
});

test('should cahnge login in state to false when changeForm is called on LoginForm', () => {
    wrapper.find('LoginForm').prop('changeForm')();
    expect(wrapper.state('login')).toBe(false);
});

// In order to make this two test work I had to import SignUpForm  query by that(I am not sure why I had to do this because two test cases above worked fine when querying only with string)
test('should call startSignUp onSubmitSignUp when I am on on SignUpForm', () => {
    wrapper.setState({ login: false });
    wrapper.find(SignUpForm).prop('onSubmit')(user);
    expect(startSignUp).toHaveBeenLastCalledWith(user);
});

test('should cahnge login in state to true when changeForm is called on SignUpForm', () => {
    wrapper.setState({ login: false });
    wrapper.find(SignUpForm).prop('changeForm')();
    expect(wrapper.state('login')).toBe(true);
});

test('should call call startLogout and removeArticles when LogginOut while on LoggedIn part', () => {
    wrapper = shallow(<UserSection 
        startSignUp={startSignUp}
        startLogin={startLogin}
        startLogout={startLogout}
        setYourArticlesFilter={setYourArticlesFilter}
        removeYourArticlesFilter={removeYourArticlesFilter}
        user={user} //<-- this is diffrent from befaoreach shallow
        authFailed={undefined}
        filter={null}
        networkError={false}
        history={history}
                    />, );
    wrapper.setState({ login: false });
    wrapper.find('LoggedIn').prop('onLogout')();
    expect(wrapper.state('login')).toBe(true);
    expect(startLogout).toHaveBeenLastCalledWith(user.token)
    expect(removeYourArticlesFilter).toHaveBeenCalledTimes(1);
    // I was not able to setup window.location.href(gosh!)
    //expect(history.push).toHaveBeenCalledWith('/');
});

test('should call setYourArticleFilter() when handleFilter is called while on LoggedIn part', () => {
    wrapper = shallow(<UserSection 
        startSignUp={startSignUp}
        startLogin={startLogin}
        startLogout={startLogout}
        setYourArticlesFilter={setYourArticlesFilter}
        removeYourArticlesFilter={removeYourArticlesFilter}
        user={user} //<-- this is diffrent from befaoreach shallow
        authFailed={undefined}
        filter={null}
        networkError={false}
        history={history}
                    />, );
    wrapper.find('LoggedIn').prop('handleFilter')();
    expect(setYourArticlesFilter).toHaveBeenCalledTimes(1);
});

test('should call removeYourArticleFilter() if filter is true and when handleFilter is called while on LoggedIn part', () => {
    wrapper = shallow(<UserSection 
        startSignUp={startSignUp}
        startLogin={startLogin}
        startLogout={startLogout}
        setYourArticlesFilter={setYourArticlesFilter}
        removeYourArticlesFilter={removeYourArticlesFilter}
        user={user} //<-- this is diffrent from befaoreach shallow
        authFailed={undefined}
        filter={true}
        networkError={false}
        history={history}
                    />, );
    wrapper.find('LoggedIn').prop('handleFilter')();
    expect(removeYourArticlesFilter).toHaveBeenCalledTimes(1);
});

test('should render Server is down when there is networkError', () => {
    wrapper = shallow(<UserSection 
        startSignUp={startSignUp}
        startLogin={startLogin}
        startLogout={startLogout}
        setYourArticlesFilter={setYourArticlesFilter}
        removeYourArticlesFilter={removeYourArticlesFilter}
        user={user} //<-- this is diffrent from befaoreach shallow
        authFailed={undefined}
        filter={null}
        networkError={true}
        history={history}
                    />, );
    expect(wrapper).toMatchSnapshot();
});
