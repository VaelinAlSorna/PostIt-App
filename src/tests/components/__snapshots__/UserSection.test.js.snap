// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Server is down when there is networkError 1`] = `
<div>
  <p>
    We are sorry but some error probably occured
  </p>
</div>
`;

exports[`should render UserSection when loginIn 1`] = `
<div>
  <LoginForm
    changeForm={[Function]}
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render UserSection when signingUp 1`] = `
<div>
  <SingUpForm
    changeForm={[Function]}
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render UserSection when user is loggedIn 1`] = `
<div>
  <LoggedIn
    filter={null}
    handleFilter={[Function]}
    name="Mark"
    onLogout={[Function]}
  />
</div>
`;
